### ISSUE TITLE
Azure Functions (node) doesn't process binary data properly

### ISSUE THREAD

**Issue by christian-vorhemus** [Azure/azure-functions-nodejs-worker#294]
My goal is to write a function which processes the binary HTTP body (for example, an image). I've created the function using the func CLI (version 2.7.2254) doing the following:

```
func init --worker-runtime "node" --language "typescript"
func new --name "SampleFunction" --template "HTTP trigger"
npm install
npm install @types/node --save
```

My function.json looks like this:
```
{
  "bindings": [
    {
      "authLevel": "function",
      "type": "httpTrigger",
      "dataType": "binary",
      "direction": "in",
      "name": "req",
      "methods": [
        "get",
        "post"
      ]
    },
    {
      "type": "http",
      "direction": "out",
      "name": "res"
    }
  ],
  "scriptFile": "../dist/SampleFunction/index.js"
}
```

My expectation would be that now req.body is a Buffer but when I check the type
`typeof req.body;`

it seems to be a string. Therefore, my attempt to save this file to disk like you see below in my index.ts also fails:

```javascript
import { AzureFunction, Context, HttpRequest } from "@azure/functions"
const fs = require('fs');

const httpTrigger: AzureFunction = async function (context: Context, req: HttpRequest): Promise<void> {
    context.log('HTTP trigger function processed a request.');
    var file = req.body;
    fs.writeFileSync("file.png", file, 'binary');
};

export default httpTrigger;
```

Test:
```
curl -X POST --data-binary "@C:\path\to\image.png" http://localhost:7
071/api/SampleHttpFunction
```

What's the correct way of processing binary data in a node based Azure Function?

**Comment by mhoeger** [comment]
"isRaw" is about the response but not the request. I also need to double check if that guidance is now deprecated, because "isRaw" was a V1 feature. 

The "dataType" property is applied by some bindings, but isn't for the HttpBinding, which is misleading :\ agreed that we should document which ones we special-case to send as binary.

**Comment by mhoeger** [comment]
@alrod and @AnatoliB - this will be super important to get into the next major version release - very much a pain today. Either by making all "body" the raw version or actually populating "rawBody" with the bytes version. [This capability](https://github.com/Azure/azure-functions-host/wiki/Out-of-Proc-Worker-Capabilities#rawhttpbodybytes) is related, most other language workers consume raw bytes for body but we couldn't do the same on this because of legacy behavior.

**Comment by midanilo** [comment]
Same problem here, I've searching a past day looking for this strange behavior on google as my binary upload doesn't have the same size and contents of my original one.
Lucky, I found this opened ticket with this undocumented tip to put content-type to application/octet-stream, but if this a known behavior by microsoft, this could be at least documented

**Comment by haukepribnow** [comment]
I also ran into this issue and was very surprised to learn about this unexpected behavior. 

After all, the current behavior is very surprising: Even if `dataType` is set to `"binary"` in the `function.json`, then `req.rawBody` still remains a `string` in any case? And `req.body` also remains a `string` unless the request is of type `application/octet-stream`? This does not feel right and does also not seem to be documented - at least not here: https://docs.microsoft.com/en-us/azure/azure-functions/functions-triggers-bindings#trigger-and-binding-definitions

In my scenario, I don't have control over the caller, i.e. I cannot enforce that my Function will get `Content-Type: application/octet-stream`. Since Proxies have been made obsolete with Functions v4, the only chance I see is to plug in a hefty API middleware - which seems to be overkill for my scenario.

I understand that backwards compatibility is important. So maybe a solution could be to introduce `req.bodyBuffer` - which is guaranteed to always contain a [`Buffer`](https://nodejs.org/api/buffer.html) object of the body.

**Comment by ejizba** [comment]
Hi folks we're working on a new programming model for Node.js and one of the benefits will be to improve the experience around backwards compatibility and breaking changes. We should be able to react much faster to requests like this, and you should have more choice when you adopt the breaking changes. See here for more details: https://github.com/Azure/azure-functions-nodejs-worker/issues/568

It's one part of our larger effort (tracked by https://github.com/Azure/azure-functions-nodejs-worker/issues/480) which is currently under way. We will go through the list of breaking changes as we're finishing up the new model and likely address a good chunk of them.

**Comment by ejizba** [comment]
FYI we took two approaches to address this issue:

1. In the existing programming model, we added a new `bufferBody` property to the http request object. The existing `body` and `rawBody` properties should behave the exact same, but the new property should always return a buffer. This has been merged, but will take a month or two to roll out in Azure
2. In the new programming model, we refactored the http request type in https://github.com/Azure/azure-functions-nodejs-library/commit/8b7588286602b28bfcffac3fd8bf4f74bb3436ef to have better naming and be closer to the fetch standard. Minor details may still change as we go through a testing/feedback phase, but the main idea should be done. I've started a doc here with more information: https://aka.ms/AzFuncNodeV4
